### ISSUE TITLE
Render Surface Normal Map?

### ISSUE THREAD

**Issue by DaoyiG** [facebookresearch/pytorch3d#865]
Hi there!

Thanks for this amazing project! I'd like to know whether it's possible to render the normal map of an object at a given pose?

Thanks!

**Comment by gkioxari** [comment]
Yes it's possible. The output from the rasterization operation is a data structure called `Fragments`

https://github.com/facebookresearch/pytorch3d/blob/4281df19cefb640067a49b961587342d9e4d85ba/pytorch3d/renderer/mesh/rasterizer.py#L16-L21

Fragments contains a variable `pix_to_face` that maps each `(x,y)` image pixel to the `K := faces_per_pixel` faces of the input mesh that are rasterized to that pixel. When rendering RGB images, that information is used to get the color of the point of intersection for each face, this is what shading and blending does. Normal rendering is even simpler. For each image pixel `(x,y)` you have the corresponding `K` faces which carry normals. So all you have to do is blend these K normal directions (if `K=1` no blending is necessary and only the nearest face is returned). 

The need to render many shape attributes (such as texture or normals) is the motivation behind breaking the computation of rendering into `rasterization - shading - blending` in PyTorch3D. The rasterization step, the hardest step in the pipeline, remains the same. But the shading and blending, which are in pure PyTorch, should be customized. In your case, you want to customize these operators to perform normal rendering instead of RGB rendering. Follow the code to see how RGB rendering is performed, and adjust your code accordingly! Note that this is the intended use of PyTorch3D.

**Comment by DaoyiG** [comment]
Thanks for your fast reply! 

If I understand you correctly, I can get the normal map by customizing a shader and the ``phong_shading`` function that returns the normals instead of the color of the K faces by something like the following:

```
def phong_normal_shading(self, meshes, fragments) -> torch.Tensor:
        faces = meshes.faces_packed()  # (F, 3)
        vertex_normals = meshes.verts_normals_packed()  # (V, 3)
        faces_normals = vertex_normals[faces]
      
        pixel_normals = interpolate_face_attributes(
            fragments.pix_to_face, fragments.bary_coords, faces_normals
        )
       
        return pixel_normals
```

The ``pixel_normals`` returned by the ``phong_normal_shading`` function will have the shape of ``(batch size * height * width * K * 3)``, then it only needs to be sent into ``softmax_rgb_blend`` function to get the blended normals (just as the blended RGB).

Is my above description correct? Please correct me if anything wrong : )

Thanks!

**Comment by gkioxari** [comment]
Very close!! Note that in `phong_shading` the `interpolate_face_attributes` will multiply the `face_attributes` (in your case the normals) with the barycentric coordinates. 

https://github.com/facebookresearch/pytorch3d/blob/9585a58d10cb2efcd159b058fa4af914203c1d0d/pytorch3d/ops/interp_face_attrs.py#L99

For RGB, that makes sense. But for normals, not so much. So your version of interpolation should just be a simper face interpolation which just averages the normals across the K faces (equivalent to passing `ones` instead of `fragments.bary_coords`.

**Comment by gkioxari** [comment]
That's right I think! If you set `K=1` then no blending is needed. For `K>1`, then you need to blend and here is where you want to make some decisions based on your needs. You might want to blend by taking the average across `K`, or you might want to blend by taking a weighted average across `K` (for example, weighing disproportionate to how far the face is from the camera -- this info is in `fragments.dist`). The `softmax_rgb_blend` does something along those lines for the latter case. If you just want to average, then you can directly call `mean` on `pixel_normals` as returned from the `interpolate_face_attributes`. 

The advantage of `soft` blending is exactly what you mention. Just taking the nearest face often leads to noisy `Fragments` especially when you learn and shapes evolve over time. So you want to propagate information to the `K` nearest faces instead of just the single *nearest* one.

**Comment by KeitoTakaishi** [comment]
I was able to get the screen space normal map using the following method.

Screen scape normal can be calculated by the gradient of the depth. 

I hope this helps you.

```
raster_settings.perspective_correct = True
fragments = rasterizer(meshes)
depth = fragments.zbuf
depth = depth[0]
depth = depth / distance #normalize

zero = torch.zeros(depth.shape).to(device)
pure_depth = depth.permute(2, 0, 1)[0, ...][None, None, ...]
depth = torch.where(depth > 0.0, depth, zero)


depth = depth.permute(2, 0, 1)[0, :, :][None, None, ...] #[b, ch, h, w]
filter = CalcNormal().to(device)
normal = filter(depth) #[b, 3, h, w]

b, c, h, w = normal.size()
one = torch.ones(size=(b, 1, h, w)).to(device)
zero = torch.zeros(size=(b, 1, h, w)).to(device)
z = torch.where(pure_depth > 0.0, one, zero)
normal[:, -1, ...] = z

class CalcNormal(nn.Module):
    def __init__(self):
        super().__init__()
        self.kernel = nn.Conv2d(in_channels=1, out_channels=3, kernel_size=3, stride=1, padding=1)
        self.kernel.state_dict()['weight'][0] = torch.FloatTensor([[0, 0, 0],
                                                                         [0.5, 0, -0.5],
                                                                         [0, 0, 0]])
        self.kernel.state_dict()['weight'][1] = torch.FloatTensor([[0, 0.5, 0],
                                                                         [0, 0, 0],
                                                                         [0, -0.5, 0]])
        self.kernel.state_dict()['weight'][2] = torch.FloatTensor([[0, 0, 0],
                                                                         [0, 0, 0],
                                                                         [0, 0, 0]])
        self.kernel.state_dict()['bias'].zero_()

    def forward(self, x):
        x = self.kernel(x)
        return x

```

![Figure_1](https://user-images.githubusercontent.com/19758089/149437477-a6a3cd10-48a7-45aa-85fb-8169888c6b0b.png)
![Figure_1](https://user-images.githubusercontent.com/19758089/149437666-ed8471c2-2843-4237-bb91-8fe79079a6bf.png)

**Comment by neverstopZyy** [comment]
> Dear @DaoyiG and @gkioxari , thank you for your sharing! However, I found in [shading](https://github.com/facebookresearch/pytorch3d/blob/main/pytorch3d/renderer/mesh/shading.py#L87) that the code pass `fragments.bary_coords` to compute the pixel normals, so why do you suggest that "passing ones instead of `fragments.bary_coords`."?

Hi dear @JasonBoy1 ,
I think the reason is the normal of all the points on a face should be the same, which is different with the utilization of color.

**Comment by yejr0229** [comment]
Hi, I wonder have you solve this? I use this code to render normal map:
![image](https://github.com/user-attachments/assets/8f99d9fb-5496-4243-88fd-05a7bfb9351b)

And I get:
![image](https://github.com/user-attachments/assets/0779d7df-2e94-40a2-a6f8-4ea64e0e83f5)

How to render a normal map like this (blue is front, red is left, green is top)?
![image](https://github.com/user-attachments/assets/bd9d0d61-7109-4370-9815-503a80def24e)

**Comment by zc-alexfan** [comment]
Custom shader for normal rendering in camera space: 

```
class NormalShader(ShaderBase):
    def __init__(self, cameras):
        super().__init__()
        self.cameras = cameras

    def forward(self, fragments, meshes, **kwargs):
        cameras = self.cameras

        faces = meshes.faces_packed()
        vertex_normals = meshes.verts_normals_packed()
        face_normals = vertex_normals[faces]

        pix_to_face = fragments.pix_to_face
        bary_coords = fragments.bary_coords
        face_normals_at_pixels = face_normals[pix_to_face]  # (N, H, W, K, 3, 3)

        interpolated_normals = (bary_coords.unsqueeze(-1) * face_normals_at_pixels).sum(dim=-2)
        interpolated_normals = torch.nn.functional.normalize(interpolated_normals, dim=-1)

        # === Transform normals to camera space
        R = cameras.R
        R = R[:, None, None, None, :, :]
        normals_cam = torch.matmul(R, interpolated_normals.unsqueeze(-1)).squeeze(-1)

        # === Fixed rotation to make normals face +Z
        theta = torch.tensor(3.1415926, device=normals_cam.device)
        cos_t = torch.cos(theta)
        sin_t = torch.sin(theta)
        R_align = torch.tensor([
            [cos_t, 0, sin_t],
            [0,     1, 0],
            [-sin_t, 0, cos_t]
        ], device=normals_cam.device).T
        normals_cam = torch.matmul(normals_cam, R_align)

        # RGB
        normal_rgb = (normals_cam + 1.0) / 2.0  # (N, H, W, K, 3)

        # Alpha: 1.0 if face index >= 0, else 0.0
        alpha = (pix_to_face >= 0).float().unsqueeze(-1)  # (N, H, W, K, 1)

        rgba = torch.cat([normal_rgb, alpha], dim=-1)  # (N, H, W, K, 4)
        return rgba
```

Also created a gist for example usage [here](https://gist.github.com/zc-alexfan/bf08c624db93adb4d02d6a94f2fadf55). 
Enjoy
